### PR TITLE
Implementation of `instanceof` operator

### LINKED ISSUE #796: Implement the `instanceof` operator
**ECMASCript feature**
We currently don't implement the `instanceof` ECMAScript operator, which would be a nice addition, since it's generating many panics in the ECMAScript test suite.

The specification of the `instanceof` operator can be found [here](https://tc39.es/ecma262/#sec-instanceofoperator).

**Example code**
Some example code can be found in [MDN](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/instanceof). But as an example:

This code should now work and give the expected result:
```javascript
let a = new String("hello");

a instanceof String;
```
The expected output is `true`.

The implementation code that needs a change is located in [here](https://github.com/boa-dev/boa/blob/master/boa/src/syntax/ast/node/operator/bin_op/mod.rs#L127-L137).

### PR DESCRIPTION
**Pull request by morrien**
- fixed @@hasinstance symbol call
- added `get_method()` and `ordinary_has_instance()` for `GcObject`
- implemented `instanceof` operator

This Pull Request fixes/closes #796.

### CODE REVIEW

File: boa/src/object/gcobject.rs [hunk 1]
```diff
@@ -676,6 +676,73 @@ impl GcObject {
...
+                        return Ok(true);
+                    }
+                    object = object_prototype.get_prototype_of();
+                }
+            
+                Ok(false)
+            } else {
+                Err(context.construct_type_error(format!(
+                    "{}",
+                    "function has non-object prototype in instanceof check"
```
Comment (HalidOdat) [inline boa/src/object/gcobject.rs:739]:
```suggestion
                    "function has non-object prototype in instanceof check",
```
We dont have to have a place holder for the main string we can just inline it

File: boa/src/syntax/ast/node/operator/bin_op/mod.rs [hunk 2]
```diff
@@ -125,15 +125,29 @@ impl Executable for BinOp {
...
+                                Some(instance_of_handler) => {
+                                    instance_of_handler.call(&y, &[x], interpreter)?.to_boolean()
+                                }
+                                None if object.is_callable() => {
+                                    object.ordinary_has_instance(interpreter, &x)?.into()
+                                }
+                                None => {
+                                    return interpreter.throw_type_error(format!(
+                                        "{}",
+                                        "right-hand side of 'instanceof' is not callable"
```
Comment (HalidOdat) [inline boa/src/syntax/ast/node/operator/bin_op/mod.rs:141]:
```suggestion
                                        "right-hand side of 'instanceof' is not callable",
```
same here

**Review by HalidOdat — COMMENTED** [review]
Looks good! this PR needs a rebase, and we should also add some tests :)

**Comment by morrien** [comment]
- fixed `throw` formatting
- added tests for `instanceof` operator and `ordinary_has_instance()`
- rebased the PR

@HalidOdat could you please explain how to resolve the remaining failing checks? Thank you.

**Comment by Razican** [comment]
>     * fixed `throw` formatting
> 
>     * added tests for `instanceof` operator and `ordinary_has_instance()`
> 
>     * rebased the PR
> 
> 
> @HalidOdat could you please explain how to resolve the remaining failing checks? Thank you.

It seems that this requires running `cargo fmt`. I will review the code, since it also seems to require a small change to fix clippy lints. The test262 issue should be solved in #914.

File: boa/src/object/gcobject.rs [hunk 3]
```diff
@@ -676,6 +676,71 @@ impl GcObject {
...
+                }
+            
+                Ok(false)
+            } else {
+                Err(context.construct_type_error(
+                    "function has non-object prototype in instanceof check"
+                ))
+            }
+        } else {
+            return Ok(false);
```
Comment (Razican) [inline boa/src/object/gcobject.rs:741]: This should solve the clippy warning:
```suggestion
            Ok(false)
```

File: boa/src/object/tests.rs [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+          function C() {}
+          C.prototype = 1
+          String instanceof C
+        } catch (err) {
+          err.toString()
+        }
+        "#;
+
+    assert_eq!(&exec(scenario), "\"TypeError: function has non-object prototype in instanceof check\"");
+}
```
Comment (Razican) [inline boa/src/object/tests.rs:16]:
```suggestion
}

```

File: boa/src/syntax/ast/node/operator/tests.rs [hunk 5]
```diff
@@ -26,3 +26,31 @@ fn assignmentoperator_rhs_throws_error() {
...
+        try {
+          let s = new String();
+          s instanceof {}
+        } catch (err) {
+          err.toString()
+        }
+        "#;
+
+    assert_eq!(&exec(scenario), "\"TypeError: right-hand side of 'instanceof' is not callable\"");
+}
```
Comment (Razican) [inline boa/src/syntax/ast/node/operator/tests.rs:56]:
```suggestion
}

```

**Review by Razican — COMMENTED** [review]
I added some suggestions on new lines at the end of files. This should be added by any editor that takes the editorconfig into account, but it seems it didn't happen in this case.

**Comment by morrien** [comment]
I ran `cargo fmt`, which also added the EOF newlines, and I fixed the clippy warning sa well.
I'll wait for [#914](https://github.com/boa-dev/boa/pull/914) to be merged and rebase the PR again.

The only failed check that remains is code coverage, how could I address that? Thank you.

**Comment by Razican** [comment]
> I ran `cargo fmt`, which also added the EOF newlines, and I fixed the clippy warning sa well.
> I'll wait for [#914](https://github.com/boa-dev/boa/pull/914) to be merged and rebase the PR again.

Thanks! That one is ready now :)

> The only failed check that remains is code coverage, how could I address that? Thank you.

We are currently debating about "patch" coverage failures. We will merge it anyways if that's the only issue.

**Comment by morrien** [comment]
Latest rebase seems to have fixed everything, even somehow improved the code coverage - I believe this is ready for merge.

With that being said, I had to rebase twice, because running `cargo clippy` would not detect another clippy warning that I noticed only in the "Files changed" tab afterwards - does anyone know what could be causing this? I believe this caused me to miss the first clippy warning regarding `return Ok(false)` as well, as I only noticed it after the PR went online.

**Comment by Razican** [comment]
> With that being said, I had to rebase twice, because running `cargo clippy` would not detect another clippy warning that I noticed only in the "Files changed" tab afterwards - does anyone know what could be causing this? I believe this caused me to miss the first clippy warning regarding `return Ok(false)` as well, as I only noticed it after the PR went online.

This might be related to https://github.com/rust-lang/rust-clippy/issues/4612

**Review by Razican — APPROVED** [review]
Looks good to me, thanks!

### Test262 conformance changes:
  | Test result | master count | PR count | difference |
  | :---------: | :----------: | :------: | :--------: |
  | Total | 78,415 | 78,415 | 0 |
  | Passed | 18,461 | 18,945 | **+484** |
  | Ignored | 15,547 | 15,547 | 0 |
  | Failed | 44,407 | 43,923 | **-484** |
  | Panics | 1,129 | 1,127 | **-2** |
  | Conformance | 23.54 | 24.16 | **+0.62%** |

**Review by RageKnify — APPROVED** [review]

**Review by HalidOdat — APPROVED** [review]

**Pull request merged**